ls1intum/Artemis
# [Exam] Enable code editor offline mode

### Checklist

### Motivation and Context
For the exam mode, the online code editor should also remain functional if the user loses his connection temporally.

### Description
1. Load repository files for exam programming exercises with online code editor enabled when the StudentExam is retrieved from the server
2. Store the participation with files in the CodeEditorFileService
3. Introduce online/offline state
4. Create wrapper checking the online/offline state. Use the server if online, otherwise use a fallback. Use the wrapper for all file operations.
5. Store file operations in the participations
6. Apply changes on the server, else sync the changes on reconnect

Known Limitation: After going offline and changing the files, once you switch to another exercise and then return, the code editor falsely indicates that your files were saved on the server. We for this PR this should be ok, but it should be fixed in a follow up PR

#### Bugfix
- [x] Exercise participation not being initialized correctly due to a LazyLoadingException being thrown when first initialising the StudentParticipation.

### Steps for Testing


_Ensure that the normal code editor works_
1. Participate in a programming exercise with online code editor enabled
2. Create, rename, delete, edit files. Saved them and commit them.
--> Everything should work as usual

_Ensure that the code editor works in exam mode if offline_
**Easier to test this on the test server**
1. Create a new exam with two programming exercises with code editor enabled. Add students, generate the student exams and participations.
2. Start the exam
3. Disconnect from the internet (if you test locally, stop the server or check if your browser has an offline mode e.g. chrome -> dev tools -> network tab -> change to offline)
4. Create, rename, delete, edit files. Switch to the other exercise. Do the same there.
--> Files and their content should be displayed correctly at all times
5. On reconnect, all requests should be sent to the server. Verify this by reloading the page a few seconds after the connection has been established again. All files should be displayed correctly.
